In [1]:
from data import CsvOHLCVLoader, MoexOHLCVLoader, BybitOHLCVLoader
from data.news import LentaNewsLoader, NewsDataFrame
from data.fundamentals import SmartLabFundamentalsLoader
from data.dataset_builder import build_dataset
import pandas as pd


**Загружаем OHLVC**

MoexLoader

In [2]:
df_moex = MoexOHLCVLoader(ticker="IMOEX", timeframe="1d", start="2025-01-01", end="2025-12-31",market= 'index',board = 'SNDX').load()
df_plzl = MoexOHLCVLoader(ticker="PLZL", timeframe="1d", start="2025-01-01", end="2025-12-31").load() 
df_sibn = MoexOHLCVLoader(ticker="SIBN", timeframe="1d", start="2025-01-01", end="2025-12-31").load()
df_sber = MoexOHLCVLoader(ticker="SBER", timeframe="1d", start="2025-01-01", end="2025-12-31").load()

In [3]:
df_sber.head()

,open,high,low,close,volume
timestamp,,,,,
2025-01-02 21:00:00+00:00,280.00,280.41,271.80,272.25,43086870
2025-01-05 21:00:00+00:00,270.88,274.41,270.07,274.37,28454750
2025-01-07 21:00:00+00:00,273.07,277.87,273.07,277.00,26634660
2025-01-08 21:00:00+00:00,276.71,278.77,270.73,271.80,52952880
2025-01-09 21:00:00+00:00,272.31,279.53,270.27,278.77,71154220


CsvLoader

In [ ]:
df_moex = CsvOHLCVLoader(path="storage/IMOEX/2020-01-01-2025-12-31.csv", start="2020-01-01", end="2025-12-31") .load()['close']
df_plzl = CsvOHLCVLoader(path="storage/PLZL/2020-01-01-2025-12-31.csv", start="2020-01-01", end="2025-12-31").load()['close']
df_sibn = CsvOHLCVLoader(path="storage/SIBN/2020-01-01-2025-12-31.csv", start="2020-01-01", end="2025-12-31") .load()['close']
df_sber = CsvOHLCVLoader(path="storage/SBER/2020-01-01-2025-12-31.csv", start="2020-01-01", end="2025-12-31") .load()['close']

In [ ]:
df_sber.head()

**Загружаем новости**

С помощью LentaLoader

In [9]:
news = LentaNewsLoader(
    start_date="2025-01-01",
    end_date="2025-12-31",
    max_pages_per_day=1,
    save=True,
    filename="storage/lenta_all_news.csv",
).load()

Collected 9369 urls


Parsing Lenta: 100%|██████████| 9369/9369 [1:12:53<00:00,  2.14it/s] 

Saved 9361 rows to storage/lenta_all_news.csv


In [ ]:
news.head()

Или читаем из CSV, чтобы не ждать 100 лет каждый раз

In [4]:
news = NewsDataFrame.read_csv('storage/lenta_all_news.csv')
news

,datetime,title,text,source,url
0,2024-06-25 00:05:00+00:00,28 июня: какой праздник отмечают в России и мире,Фото: Андрей Бок / РИА Новости 28 июня в Петер...,Lenta.ru,https://lenta.ru/news/2024/06/25/28-iyunya-kak...
1,2024-08-30 00:05:00+00:00,30 августа: какой праздник отмечают в России и...,Фото: Максим Богодвид / РИА Новости 30 августа...,Lenta.ru,https://lenta.ru/news/2024/08/30/30-avgusta-ka...
2,2024-11-15 00:05:00+00:00,15 ноября: какой праздник отмечают в России и ...,Фото: beeboys / Shutterstock / Fotodom 15 нояб...,Lenta.ru,https://lenta.ru/news/2024/11/15/15-noyabrya-k...
3,2024-12-31 00:03:00+00:00,Протестующие в Тбилиси накрыли праздничный стол,Фото: Daro Sulakauri / Reuters Протестующие на...,Lenta.ru,https://lenta.ru/news/2024/12/31/protestuyusch...
4,2024-12-31 00:23:00+00:00,Моряки с танкера Eagle S попали под следствие ...,Фото: Jussi Nukari / Reuters Семь моряков с та...,Lenta.ru,https://lenta.ru/news/2024/12/31/sem-moryakov-...
...,...,...,...,...,...
9356,2025-12-31 01:45:00+00:00,Экс-аналитик ЦРУ заявил о развале НАТО,Конфликт на Украине и налаживание отношений ме...,Lenta.ru,https://lenta.ru/news/2025/12/31/usa/
9357,2025-12-31 01:56:00+00:00,Евросоюзу напомнили о соседстве с Россией,Директор департамента европейских проблем МИД ...,Lenta.ru,https://lenta.ru/news/2025/12/31/evrosoyuzu-na...
9358,2025-12-31 02:02:00+00:00,В США рассказали о проблемах программы F-35,Фото: jgorzynik / Shutterstock / Fotodom Пента...,Lenta.ru,https://lenta.ru/news/2025/12/31/v-ssha-rasska...
9359,2025-12-31 02:04:00+00:00,В Подмосковье произошла массовая драка со стре...,Фото: Алексей Сухоруков / РИА Новости В подмос...,Lenta.ru,https://lenta.ru/news/2025/12/31/v-podmoskovie...


Отфильтруем новости по тикеру и отрасли

*Полюс*

In [5]:
plzl_news = news.keywords(["PLZL", "Полюс", "акции полюса","акции Полюса",'золото','золотодобывающая','Сухой Лог','Золото'])
plzl_news.save('storage/PLZL.csv')

Saved 92 rows to storage/PLZL.csv


*Сбер*

In [6]:
sber_news = news.keywords(["SBER", "Сбер", "акции сбербанка","акции сбера",'банк','банки','банкоматы','GigaChat','gigachat','Gigachat'])
sber_news.save('storage/SBER.csv')

Saved 265 rows to storage/SBER.csv


*Сибнефть*

In [7]:
sibn_news = news.keywords(["SIBN", "Сибнефть", "акции сибнефти","акции сибинтек",'нефть','нефтепереработка','нефтяная','Газпром нефть','газпром нефть'])
sibn_news.save('storage/SIBN.csv')

Saved 58 rows to storage/SIBN.csv


*Моекс (индекс)*

In [8]:
imoex_news = news.keywords(["IMOEX", "Moex", "индекс Мосбиржи","индекс мосбиржи",'Моекс','моекс','имоекс','мосбиржи',' фондовый индекс'])
imoex_news.save('storage/IMOEX.csv')

Saved 2 rows to storage/IMOEX.csv


**Загружаем фундаменталы**

In [9]:
fundamentals = SmartLabFundamentalsLoader(
    tickers=["SBER", "PLZL", "SIBN"],
    metrics=[
        "p_e",
        "roe",
        "net_income"
    ],
    start="2025-01-01",
    end="2025-12-31",
    period_type="quarter",
    report_type="MSFO",
    save=True,
    filename="storage/fundamentals_quarterly.csv",
).load()

fundamentals


Ticker: SBER
  p_e: 3 rows
  roe: 3 rows
  net_income: 3 rows

Ticker: PLZL
  p_e: 3 rows
  roe: 4 rows
  net_income: 4 rows

Ticker: SIBN
  p_e: 3 rows
  roe: 3 rows
  net_income: 3 rows
Saved 10 rows to storage/fundamentals_quarterly.csv


,ticker,period,period_date,available_date,pe,roe,net_income
0,PLZL,2024Q4,2024-12-31,2025-01-31,NaN,241.1,175.4
1,PLZL,2025Q1,2025-03-31,2025-04-30,8.29,NaN,NaN
2,PLZL,2025Q2,2025-06-30,2025-07-31,7.16,149.8,172.5
3,PLZL,2025Q3,2025-09-30,2025-10-31,9.36,NaN,NaN
4,SBER,2025Q1,2025-03-31,2025-04-30,4.33,24.4,436.0
5,SBER,2025Q2,2025-06-30,2025-07-31,4.41,23.0,422.6
6,SBER,2025Q3,2025-09-30,2025-10-31,3.91,23.7,449.8
7,SIBN,2025Q1,2025-03-31,2025-04-30,6.66,13.2,92.6
8,SIBN,2025Q2,2025-06-30,2025-07-31,8.56,9.9,57.8
9,SIBN,2025Q3,2025-09-30,2025-10-31,10.80,7.2,69.1


**Мерджим таблицы**

In [10]:
items = {
    "plzl": (df_plzl, plzl_news, "PLZL"),
    "sber": (df_sber, sber_news, "SBER"),
    "sibn": (df_sibn, sibn_news, "SIBN"),
}

final = {}

for name, (ohlcv_df, news_df, ticker) in items.items():
    final[name] = build_dataset(
        ohlcv=ohlcv_df,
        news=news_df,
        fundamentals=fundamentals,
        ticker=ticker,
        drop_nans=True,
    )

plzl_final = final["plzl"]
sber_final = final["sber"]
sibn_final = final["sibn"]

In [11]:
sber_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 223 entries, 0 to 222
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype              
---  ------          --------------  -----              
 0   timestamp       223 non-null    datetime64[ns, UTC]
 1   open            223 non-null    float64            
 2   high            223 non-null    float64            
 3   low             223 non-null    float64            
 4   close           223 non-null    float64            
 5   volume          223 non-null    int64              
 6   date            223 non-null    object             
 7   news            223 non-null    object             
 8   period          223 non-null    object             
 9   period_date     223 non-null    datetime64[ns]     
 10  available_date  223 non-null    datetime64[ns, UTC]
 11  pe              223 non-null    float64            
 12  roe             223 non-null    float64            
 13  net_income      223 non-null    flo

In [16]:
sber_final.head()

,timestamp,open,high,low,close,volume,date,news,period,period_date,available_date,pe,roe,net_income,ticker
0,2025-05-01 21:00:00+00:00,307.80,307.80,299.70,299.96,19751770,2025-05-01,[Фото: Maksim Konstantinov / Globallookpress.c...,2025Q1,2025-03-31,2025-04-30 00:00:00+00:00,4.33,24.4,436.0,SBER
1,2025-05-02 21:00:00+00:00,299.90,302.24,299.52,300.70,2942530,2025-05-02,[С начала 2025 года рубль сильнее других валют...,2025Q1,2025-03-31,2025-04-30 00:00:00+00:00,4.33,24.4,436.0,SBER
2,2025-05-03 21:00:00+00:00,300.70,302.78,300.02,301.80,2867700,2025-05-03,[],2025Q1,2025-03-31,2025-04-30 00:00:00+00:00,4.33,24.4,436.0,SBER
3,2025-05-04 21:00:00+00:00,301.80,302.59,291.66,294.64,45508060,2025-05-04,[Фото: Алексей Никольский / РИА Новости Глобал...,2025Q1,2025-03-31,2025-04-30 00:00:00+00:00,4.33,24.4,436.0,SBER
4,2025-05-05 21:00:00+00:00,294.65,304.75,292.05,302.02,49327640,2025-05-05,[],2025Q1,2025-03-31,2025-04-30 00:00:00+00:00,4.33,24.4,436.0,SBER


In [13]:
plzl_final[["timestamp", "ticker", "period", "available_date", "pe", "roe", "net_income"]].head(20)

,timestamp,ticker,period,available_date,pe,roe,net_income
0,2025-02-02 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
1,2025-02-03 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
2,2025-02-04 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
3,2025-02-05 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
4,2025-02-06 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
5,2025-02-09 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
6,2025-02-10 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
7,2025-02-11 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
8,2025-02-12 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4
9,2025-02-13 21:00:00+00:00,PLZL,2024Q4,2025-01-31 00:00:00+00:00,NaN,241.1,175.4


**Все, все данные есть. Теперь начинаем исследование. В дальнейшей работе я беру датасет СБЕРА, потому что у сбера больше новостей 265**

Оценим каждую новость по сантименту и сделаем из этого фактор: 1 - положительная; 0 - отрицательная